rabbit/src/lib.rs

```diff
@@ -0,0 +1,515 @@
+    state.counter_vars[7] = state.counter_vars[7] ^ i3;
+
+    for _ in 0..4 {
+        next_state(state);
+    }
+}
+
+/// RFC 4503. 2.5.  Counter System (page 3).
+fn counter_update(state: &mut State) {
+    for j in 0..8 {
+        let temp = state.counter_vars[j] as u64 + A[j] as u64 + state.carry_bit as u64;
+        state.carry_bit = ((temp / WORDSIZE) as u8) & 0b1;
+        state.counter_vars[j] = (temp % WORDSIZE) as u32;
+    }
+}
+
+/// RFC 4503. 2.6.  Next-State Function (page 3-4).
+fn next_state(state: &mut State) {
+    let mut g = [0u32; 8];
+
+    counter_update(state);
+
+    for j in 0..8 {
+        let u_plus_v = state.state_vars[j] as u64 + state.counter_vars[j] as u64;
+        let square_uv = (u_plus_v % WORDSIZE) * (u_plus_v % WORDSIZE);
+        g[j] = (square_uv ^ (square_uv >> 32)) as u32;
+    }
+
+    state.state_vars[0] = g[0]
+        .wrapping_add(g[7].rotate_left(16))
+        .wrapping_add(g[6].rotate_left(16));
+    state.state_vars[1] = g[1].wrapping_add(g[0].rotate_left(8)).wrapping_add(g[7]);
+    state.state_vars[2] = g[2]
+        .wrapping_add(g[1].rotate_left(16))
+        .wrapping_add(g[0].rotate_left(16));
+    state.state_vars[3] = g[3].wrapping_add(g[2].rotate_left(8)).wrapping_add(g[1]);
+    state.state_vars[4] = g[4]
+        .wrapping_add(g[3].rotate_left(16))
+        .wrapping_add(g[2].rotate_left(16));
+    state.state_vars[5] = g[5].wrapping_add(g[4].rotate_left(8)).wrapping_add(g[3]);
+    state.state_vars[6] = g[6]
+        .wrapping_add(g[5].rotate_left(16))
+        .wrapping_add(g[4].rotate_left(16));
+    state.state_vars[7] = g[7].wrapping_add(g[6].rotate_left(8)).wrapping_add(g[5]);
+}
+
+/// RFC 4503. 2.7.  Extraction Scheme (page 4).
+fn extract(state: &State) -> [u8; 16] {
+    let mut s = [0u8; 16];
+
+    let s15_0 = ((state.state_vars[0]) ^ (state.state_vars[5] >> 16)) as u16;
+    let s31_16 = ((state.state_vars[0] >> 16) ^ (state.state_vars[3])) as u16;
+    let s47_32 = ((state.state_vars[2]) ^ (state.state_vars[7] >> 16)) as u16;
+    let s63_48 = ((state.state_vars[2] >> 16) ^ (state.state_vars[5])) as u16;
+    let s79_64 = ((state.state_vars[4]) ^ (state.state_vars[1] >> 16)) as u16;
+    let s95_80 = ((state.state_vars[4] >> 16) ^ (state.state_vars[7])) as u16;
+    let s111_96 = ((state.state_vars[6]) ^ (state.state_vars[3] >> 16)) as u16;
+    let s127_112 = ((state.state_vars[6] >> 16) ^ (state.state_vars[1])) as u16;
+
+    s[0x0] = s15_0 as u8;
+    s[0x1] = (s15_0 >> 8) as u8;
+    s[0x2] = s31_16 as u8;
+    s[0x3] = (s31_16 >> 8) as u8;
+    s[0x4] = s47_32 as u8;
+    s[0x5] = (s47_32 >> 8) as u8;
+    s[0x6] = s63_48 as u8;
+    s[0x7] = (s63_48 >> 8) as u8;
+    s[0x8] = s79_64 as u8;
+    s[0x9] = (s79_64 >> 8) as u8;
+    s[0xA] = s95_80 as u8;
+    s[0xB] = (s95_80 >> 8) as u8;
+    s[0xC] = s111_96 as u8;
+    s[0xD] = (s111_96 >> 8) as u8;
+    s[0xE] = s127_112 as u8;
+    s[0xF] = (s127_112 >> 8) as u8;
+
+    return s;
```

Comment:
To fix the clippy failure, change this to:

```suggestion
    s
```